src/java/org/apache/cassandra/auth/AuthenticatedUser.java

```diff
@@ -33,10 +35,10 @@
 public class AuthenticatedUser
 {
     public static final String SYSTEM_USERNAME = "system";
-    public static final AuthenticatedUser SYSTEM_USER = new AuthenticatedUser(SYSTEM_USERNAME);
+    public static final AuthenticatedUser SYSTEM_USER = new AuthenticatedUser(SYSTEM_USERNAME, SYSTEM_USERNAME);
```

Comment:
again, using same constant for "authentication mode" just feels strange. What has _username_ common with _authentication mode_?

Reply:
I had considered an Enum, but one thing I was mindful of was third party `IAuthenticator` implementations.  If they wanted to introduce their own mode of auth, they would not be able to without making a change to the core codebase since Enums are effectively final.  

Is this something we should support?  If not I'm fine with using an Enum.

Reply:
🤔 so use an Enum internally, and `String` on the public API methods?  That seems like a good compromise, i'll give that a try?

Reply:
Just pushed [605d9cb](https://github.com/apache/cassandra/pull/3085/commits/605d9cbaa047c79ba7f822a5c9c4ea915de7f269) which includes `AuthenticationMode` enum in `IAuthenticator`.

It feels a bit weird that i'm converting the enum to a String in a number of places, seems like it could be error prone and I feel like we're not getting the full value of using an enum.

I had a few more possible ideas I'd throw out there to gauge your thoughts:

1. Add an enum value `CUSTOM` to `AuthenticationMode`, that 3rd party implementors can use to indicate that the authentication mode used is something 'custom'.    This would allow us to use `AuthenticationMode` all throughout the API (without having to convert explicitly to String anywhere except maybe in the system table).  The main downside is that it's not very specific, but I think for most implementors this would be sufficient as I don't expect an implementor will support multiple custom modes in their authenticator.
2. Instead of using an enum, introduce a class for `AuthenticationMode` that a third party could implement. I feel this one offers less value, and is only a slight improvement over string.

option 1 isn't perfect but I think could be an improvement over what I just implemented and will make the API cleaner and less error prone.  What do you think?

Reply:
I think the approach 2) is a little bit superior here.  The reason I think so is that, as you said, CUSTOM is quite non-telling and it would also mean that absolutely every other authentication mode, not part of core Cassandra, would be CUSTOM ever. I don't like this. In system_views.clients, it would be all "Custom", for JMX as well ... We can definitely do better. It would make people question what actual authenticator that node is configured with if they do not see something explicit but just "Custom".

I wanted to see how it would look like on a class so I implemented it here https://github.com/tolbertam/cassandra/pull/1

It is just a commit on top of your latest stuff here.

Reply:
Thanks for implementing that, that looks very good and does look like a much better solution.  I'll cherry pick it over, thank you!